File: examples/typescript/r2/r2.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import dotenv from 'dotenv'
import { MountMode, R2Resource, Workspace, type FileStat, type R2Config } from '@struktoai/mirage-node'

dotenv.config({ path: '.env.development' })

function configFromEnv(): R2Config {
  const bucket = process.env.R2_BUCKET
  const accessKeyId = process.env.R2_ACCESS_KEY_ID
  const secretAccessKey = process.env.R2_SECRET_ACCESS_KEY
  if (bucket === undefined || accessKeyId === undefined || secretAccessKey === undefined) {
    throw new Error(
      'R2_BUCKET, R2_ACCESS_KEY_ID, R2_SECRET_ACCESS_KEY must be set (e.g. in .env.development); also provide R2_ACCOUNT_ID or R2_ENDPOINT_URL',
    )
  }
  const accountId = process.env.R2_ACCOUNT_ID
  const endpoint = process.env.R2_ENDPOINT_URL
  const region = process.env.R2_REGION
  if (accountId === undefined && endpoint === undefined) {
    throw new Error('R2 requires R2_ACCOUNT_ID or R2_ENDPOINT_URL to be set')
  }
  return {
    bucket,
    accessKeyId,
    secretAccessKey,
    ...(accountId !== undefined ? { accountId } : {}),
    ...(endpoint !== undefined ? { endpoint } : {}),
    ...(region !== undefined ? { region } : {}),
  }
}

async function run(
  ws: Workspace,
  command: string,
): Promise<{ stdout: string; stderr: string; exitCode: number }> {
  const r = await ws.execute(command)
  return {
    stdout: r.stdoutText,
    stderr: r.stderrText,
    exitCode: r.exitCode,
  }
}

async function main(): Promise<void> {
  const config = configFromEnv()
  const resource = new R2Resource(config)
  const ws = new Workspace({ '/r2/': resource }, { mode: MountMode.READ })

  try {
    console.log('=== ls /r2/ ===')
    console.log((await run(ws, 'ls /r2/')).stdout)

    console.log('=== ls /r2/data/ ===')
    console.log((await run(ws, 'ls /r2/data/')).stdout)

    console.log('=== tree /r2/ ===')
    console.log((await run(ws, 'tree /r2/')).stdout)

    console.log('=== stat /r2/data (directory prefix) ===')
    console.log(`  ${(await run(ws, 'stat /r2/data')).stdout.trim()}`)

    console.log('\n=== stat /r2/data/example.json ===')
    console.log(`  ${(await run(ws, 'stat /r2/data/example.json')).stdout.trim()}`)

    console.log('\n=== cat /r2/data/example.json | head -n 10 ===')
    console.log((await run(ws, 'cat /r2/data/example.json | head -n 10')).stdout)

    console.log('=== head -n 3 /r2/data/example.jsonl ===')
    console.log((await run(ws, 'head -n 3 /r2/data/example.jsonl')).stdout.slice(0, 300))

    console.log('\n=== tail -n 2 /r2/data/example.jsonl ===')
    console.log((await run(ws, 'tail -n 2 /r2/data/example.jsonl')).stdout.slice(0, 300))

    console.log('\n=== wc -l /r2/data/example.jsonl ===')
    console.log(`  ${(await run(ws, 'wc -l /r2/data/example.jsonl')).stdout.trim()}`)

    console.log('\n=== grep -c mirage /r2/data/example.jsonl ===')
    console.log(`  count: ${(await run(ws, 'grep -c mirage /r2/data/example.jsonl')).stdout.trim()}`)

    console.log('\n=== grep mirage /r2/data/example.jsonl | head -n 3 ===')
    const grepOut = (await run(ws, 'grep mirage /r2/data/example.jsonl | head -n 3')).stdout
    for (const ln of grepOut.trim().split('\n')) console.log(`  ${ln.slice(0, 100)}...`)

    console.log("\n=== find /r2/ -name '*.json' ===")
    console.log((await run(ws, "find /r2/ -name '*.json'")).stdout)

    console.log("=== find /r2/ -name '*.parquet' ===")
    console.log((await run(ws, "find /r2/ -name '*.parquet'")).stdout)

    console.log('=== jq .metadata /r2/data/example.json ===')
    console.log(`  ${(await run(ws, 'jq .metadata /r2/data/example.json')).stdout.trim().slice(0, 200)}`)

    console.log("\n=== jq '.departments[].teams[].name' /r2/data/example.json ===")
    console.log(
      `  ${(await run(ws, 'jq ".departments[].teams[].name" /r2/data/example.json')).stdout.trim()}`,
    )

    console.log('\n=== cat example.jsonl | grep queue-operation | sort | uniq | wc -l ===')
    console.log(
      `  unique lines: ${(await run(ws, 'cat /r2/data/example.jsonl | grep queue-operation | sort | uniq | wc -l')).stdout.trim()}`,
    )

    console.log('\n=== pwd ===')
    console.log(`  ${(await run(ws, 'pwd')).stdout.trim()}`)

    console.log('\n=== cd /r2/data ===')
    console.log(`  exit=${(await run(ws, 'cd /r2/data')).exitCode}`)

    console.log('\n=== pwd (after cd) ===')
    console.log(`  ${(await run(ws, 'pwd')).stdout.trim()}`)

    console.log('\n=== ls (relative) ===')
    console.log((await run(ws, 'ls')).stdout)

    console.log('=== head -n 3 example.json (relative) ===')
    console.log((await run(ws, 'head -n 3 example.json')).stdout)

    console.log('\n=== cat | grep | head (streaming drain) ===')
    const streamed = (await run(ws, 'cat /r2/data/example.jsonl | grep queue | head -n 3')).stdout
    console.log(`  got ${streamed.trim().split('\n').length} lines (expected 3)`)

    console.log('\n=== grep -q && echo (barrier VALUE) ===')
    let r = await run(ws, 'grep -q queue /r2/data/example.jsonl && echo "found"')
    console.log(`  stdout: ${r.stdout.trim()}`)
    console.log(`  exit: ${r.exitCode}`)

    console.log('\n=== grep -q || echo (barrier OR) ===')
    r = await run(ws, 'grep -q NONEXISTENT_STRING /r2/data/example.jsonl || echo "not found"')
    console.log(`  stdout: ${r.stdout.trim()}`)
    console.log(`  exit: ${r.exitCode}`)

    console.log('\n=== grep ; grep (semicolon materialization) ===')
    r = await run(
      ws,
      'grep -c queue /r2/data/example.jsonl; grep -c mirage /r2/data/example.jsonl',
    )
    console.log(`  stdout: ${r.stdout.trim()}`)

    console.log('\n=== grep missing ; echo $? (semicolon exit code) ===')
    r = await run(ws, 'grep NONEXISTENT_STRING /r2/data/example.jsonl; echo $?')
    console.log(`  stdout: ${r.stdout.trim()}`)

    console.log('\n=== cat nonexistent 2>&1 | head (stderr in pipe) ===')
    r = await run(ws, 'cat /r2/data/nonexistent_file 2>&1 | head -n 1')
    console.log(`  stdout: ${r.stdout.trim()}`)
    console.log(`  exit: ${r.exitCode}`)

    console.log('\n=== cat nonexistent | cat (no merge: error in stderr) ===')
    r = await run(ws, 'cat /r2/data/nonexistent_file | cat')
    console.log(`  stdout: '${r.stdout.trim()}' (expect empty)`)
    console.log(`  stderr: '${r.stderr.trim().slice(0, 80)}'`)

    console.log('\n=== cat nonexistent 2>&1 | cat (no double-emit) ===')
    r = await run(ws, 'cat /r2/data/nonexistent_file 2>&1 | cat')
    console.log(`  stdout: '${r.stdout.trim().slice(0, 80)}'`)
    console.log(`  stderr: '${r.stderr.trim()}' (expect empty: no double-emit)`)

    console.log('\n=== cat large 2>&1 | wc -l (streams real payload) ===')
    const wcRes = await run(ws, 'wc -l /r2/data/example.jsonl')
    const expected = Number.parseInt(wcRes.stdout.trim().split(/\s+/)[0] ?? '0', 10)
    const mergedRes = await run(ws, 'cat /r2/data/example.jsonl 2>&1 | wc -l')
    const got = Number.parseInt(mergedRes.stdout.trim(), 10)
    console.log(`  expected: ${expected}  got: ${got}  ${got === expected ? 'OK' : 'MISMATCH'}`)

    console.log('\n=== cat | sort | uniq | wc -l (full pipeline) ===')
    r = await run(ws, 'cat /r2/data/example.jsonl | sort | uniq | wc -l')
    console.log(`  unique lines: ${r.stdout.trim()}`)

    console.log('\n=== grep -c & echo kicked off; wait (bg job) ===')
    r = await run(ws, "grep -c queue /r2/data/example.jsonl & echo 'kicked off'; wait")
    console.log(`  stdout: ${r.stdout.trim()}`)

    console.log("\n=== sleep 0 & cat (bg doesn't consume stdin) ===")
    r = await run(ws, 'sleep 0 & cat /r2/data/example.json | head -n 1')
    console.log(`  stdout: ${r.stdout.trim()}`)

    console.log('\n=== cat nonexistent & echo ok (bg error handled) ===')
    r = await run(ws, 'cat /r2/data/nonexistent_file & echo ok; wait; echo done')
    console.log(`  stdout: ${r.stdout.trim()}`)
    console.log(`  exit: ${r.exitCode}`)

    console.log('\n=== multiple bg: grep & wc & wait (parallel) ===')
    r = await run(
      ws,
      'grep -c queue /r2/data/example.jsonl & wc -l /r2/data/example.jsonl & wait; echo all done',
    )
    console.log(`  stdout: ${r.stdout.trim()}`)

    console.log('\n=== head -n 5 | while read; do echo (bounded loop) ===')
    r = await run(
      ws,
      'cat /r2/data/example.jsonl | head -n 5 | while read LINE; do echo got; done | wc -l',
    )
    console.log(`  iterations: ${r.stdout.trim()} (expected 5)`)

    console.log('\n=== while read; break (early exit) ===')
    r = await run(
      ws,
      'cat /r2/data/example.jsonl | head -n 100 | while read LINE; do echo first; break; done',
    )
    const lines = r.stdout.trim().split('\n')
    console.log(`  stdout lines: ${lines.length} (expected 1)  exit=${r.exitCode}`)

    console.log('\n=== for x in a b c; do read LINE (loop reads buffer) ===')
    r = await run(
      ws,
      'cat /r2/data/example.jsonl | head -n 3 | for x in a b c; do read LINE; echo "$x:${LINE:0:30}"; done',
    )
    for (const ln of r.stdout.trim().split('\n')) console.log(`  ${ln}`)

    console.log('\n=== echo "\\$X" (escaped dollar stays literal) ===')
    await run(ws, 'export X=expanded')
    r = await run(ws, 'echo "\\$X"')
    console.log(`  stdout: ${JSON.stringify(r.stdout.trim())} (expect '$X')`)

    console.log('\n=== echo "$X" (unescaped dollar expands) ===')
    r = await run(ws, 'echo "$X"')
    console.log(`  stdout: ${JSON.stringify(r.stdout.trim())} (expect 'expanded')`)

    console.log("\n=== echo '$X' (single quotes keep $X literal) ===")
    r = await run(ws, "echo '$X'")
    console.log(`  stdout: ${JSON.stringify(r.stdout.trim())} (expect '$X')`)

    console.log('\n=== cat "$DIR/example.json" (env var in path) ===')
    await run(ws, 'export DIR=/r2/data')
    r = await run(ws, 'cat "$DIR/example.json" | head -n 3')
    console.log(`  first lines: ${JSON.stringify(r.stdout.trim().split('\n'))}`)

    console.log('\n=== cat $(echo /r2/data/example.json) | head -n 1 (command sub as path) ===')
    r = await run(ws, 'cat $(echo /r2/data/example.json) | head -n 1')
    console.log(`  stdout: ${r.stdout.trim()}`)

    console.log('\n=== grep "$(echo queue)" /r2/data/example.jsonl | wc -l (sub as pattern) ===')
    r = await run(ws, 'grep "$(echo queue)" /r2/data/example.jsonl | wc -l')
    console.log(`  count: ${r.stdout.trim()}`)


    // chmod/chown/touch never hit the R2 API: attrs land in the
    // workspace namespace (durable, snapshot-captured) and merge into
    // dispatch-level stat.
    console.log(`=== metadata overlay on /r2/data/example.jsonl ===`)
    const metaRes = await ws.execute(
      `chmod 640 "/r2/data/example.jsonl" && chown 500:dev "/r2/data/example.jsonl" && touch -t 202601021530 "/r2/data/example.jsonl"`,
    )
    console.log(`  chmod/chown/touch exit=${String(metaRes.exitCode)}`)
    const metaSt = (await ws.dispatch('stat', `/r2/data/example.jsonl`)) as FileStat
    const metaMode = metaSt.mode !== null ? metaSt.mode.toString(8) : '-'
    console.log(
      `  dispatch stat: mode=${metaMode} uid=${String(metaSt.uid)} gid=${String(metaSt.gid)} mtime=${String(metaSt.modified)}`,
    )
  } finally {
    await ws.close()
  }
}

main().catch((err: unknown) => {
  console.error(err)
  process.exit(1)
})

File: examples/python/r2/r2.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import asyncio
import json
import os

from dotenv import load_dotenv

from mirage import MountMode, Workspace
from mirage.resource.r2 import R2Config, R2Resource
from mirage.types import PathSpec

load_dotenv(".env.development")

config = R2Config(
    bucket=os.environ["R2_BUCKET"],
    account_id=os.environ.get("R2_ACCOUNT_ID"),
    endpoint_url=os.environ.get("R2_ENDPOINT_URL"),
    access_key_id=os.environ["R2_ACCESS_KEY_ID"],
    secret_access_key=os.environ["R2_SECRET_ACCESS_KEY"],
    region=os.environ.get("R2_REGION", "auto"),
)

backend = R2Resource(config)
ws = Workspace({"/r2/": backend}, mode=MountMode.READ)


def ops_summary() -> str:
    records = ws.ops.records
    total = sum(r.bytes for r in records)
    return f"{len(records)} ops, {total} bytes transferred"


async def main():
    print("=== PLAN ESTIMATES ===\n")

    dr = await ws.execute("grep mirage /r2/data/example.jsonl", provision=True)
    print("--- plan: grep mirage /r2/data/example.jsonl ---")
    print(f"  network_read: {dr.network_read}, cache_read: {dr.cache_read}")
    print(f"  read_ops: {dr.read_ops}, precision: {dr.precision}")

    dr = await ws.execute("grep mirage /r2/data/example.jsonl | head -n 3",
                          provision=True)
    print("\n--- plan: grep mirage ... | head -n 3 ---")
    print(f"  op: {dr.op}, children: {len(dr.children)}")
    print(f"  network_read: {dr.network_read}, cache_read: {dr.cache_read}")
    print(f"  precision: {dr.precision}")
    for c in dr.children:
        net, cache = c.network_read, c.cache_read
        print(f"    {c.command}: net={net}, cache={cache}, {c.precision}")

    dr = await ws.execute("grep mirage /r2/data/example.jsonl && echo found",
                          provision=True)
    print("\n--- plan: grep ... && echo found ---")
    print(f"  op: {dr.op}, network_read: {dr.network_read}")
    for c in dr.children:
        print(f"    {c.command}: net={c.network_read}, {c.precision}")

    print(f"\n  Stats after plans (should be 0): {ops_summary()}")

    print("\n--- caching: cat /r2/data/example.jsonl | wc -l ---")
    result = await ws.execute("cat /r2/data/example.jsonl | wc -l")
    print(f"  lines: {(await result.stdout_str()).strip()}")
    print(f"  Stats after caching: {ops_summary()}")

    dr = await ws.execute("grep mirage /r2/data/example.jsonl", provision=True)
    print("\n--- plan after cache: grep mirage ... ---")
    print(f"  network_read: {dr.network_read}, cache_read: {dr.cache_read}")
    print(f"  cache_hits: {dr.cache_hits}, read_ops: {dr.read_ops}")

    print("\n=== ACTUAL EXECUTION ===\n")

    print("--- grep mirage /r2/data/example.jsonl ---")
    output = await (
        await ws.execute("grep mirage /r2/data/example.jsonl")).stdout_str()
    lines = output.strip().splitlines() if output.strip() else []
    print(f"  Matches: {len(lines)}")
    if lines:
        print(f"  First: {lines[0][:80]}...")
    print(f"  Stats: {ops_summary()}")

    print("\n--- grep -m 1 mirage /r2/data/example.jsonl ---")
    output = await (
        await
        ws.execute("grep -m 1 mirage /r2/data/example.jsonl")).stdout_str()
    lines = output.strip().splitlines() if output.strip() else []
    print(f"  Matches: {len(lines)}")
    print(f"  Stats: {ops_summary()}")

    print("\n--- grep mirage /r2/data/example.jsonl | wc -l ---")
    result = await ws.execute("grep mirage /r2/data/example.jsonl | wc -l")
    print(f"  Count: {(await result.stdout_str()).strip()}")
    print(f"  Exit code: {result.exit_code}")
    print(f"  Stats: {ops_summary()}")

    print("\n--- grep mirage /r2/data/example.jsonl | head -n 3 ---")
    result = await ws.execute("grep mirage /r2/data/example.jsonl | head -n 3")
    lines = (await result.stdout_str()).strip().splitlines()
    print(f"  Lines: {len(lines)}")
    for ln in lines:
        print(f"    {ln[:80]}...")
    print(f"  Stats: {ops_summary()}")

    print("\n--- cat /r2/data/example.jsonl"
          " | grep queue-operation | sort | uniq ---")
    result = await ws.execute(
        "cat /r2/data/example.jsonl | grep queue-operation | sort | uniq")
    lines = ((await result.stdout_str()).strip().splitlines() if
             (await result.stdout_str()).strip() else [])
    print(f"  Unique lines: {len(lines)}")
    print(f"  Stats: {ops_summary()}")

    print("\n--- rg queue-operation /r2/data/example.jsonl"
          " | head -n 5 | cut -d , -f 2 ---")
    result = await ws.execute(
        "rg queue-operation /r2/data/example.jsonl | head -n 5 | cut -d , -f 2"
    )
    print(f"  Fields:\n    {(await result.stdout_str()).strip()}")
    print(f"  Stats: {ops_summary()}")

    print("\n--- grep -m 1 mirage /r2/data/example.jsonl"
          " && echo 'found mirage' ---")
    result = await ws.execute(
        "grep -m 1 mirage /r2/data/example.jsonl && echo found")
    print(f"  Exit code: {result.exit_code}")
    print(
        f"  Stdout ends with: ...{(await result.stdout_str()).strip()[-30:]}")
    print(f"  Stats: {ops_summary()}")

    print("\n--- grep NONEXISTENT /r2/data/example.jsonl"
          " || echo 'not found' ---")
    result = await ws.execute(
        "grep NONEXISTENT /r2/data/example.jsonl || echo not_found")
    print(f"  Exit code: {result.exit_code}")
    print(f"  Output: {(await result.stdout_str()).strip()}")
    print(f"  Stats: {ops_summary()}")

    print("\n--- (grep queue-operation /r2/data/example.jsonl"
          " | sort | uniq) | wc -l ---")
    result = await ws.execute(
        "(grep queue-operation /r2/data/example.jsonl | sort | uniq) | wc -l")
    print(f"  Unique queue ops: {(await result.stdout_str()).strip()}")
    print(f"  Stats: {ops_summary()}")

    print("\n--- head -n 1 /r2/data/example.jsonl"
          " ; wc -l /r2/data/example.jsonl ---")
    result = await ws.execute(
        "head -n 1 /r2/data/example.jsonl ; wc -l /r2/data/example.jsonl")
    print(f"  Output: {(await result.stdout_str()).strip()}")
    print(f"  Stats: {ops_summary()}")

    print("\n--- lazy multi-pipe: grep | grep -v | head | cut ---")
    result = await ws.execute("grep queue-operation /r2/data/example.jsonl"
                              " | grep -v error | head -n 2 | cut -d , -f 1")
    print(f"  Output:\n    {(await result.stdout_str()).strip()}")

    result_full = await ws.execute(
        "grep queue-operation /r2/data/example.jsonl"
        " | grep -v error | cut -d , -f 1")
    full_lines = (await result_full.stdout_str()).strip().splitlines()
    print(f"  Without head: {len(full_lines)} lines (full R2 download)")

    print("\n--- rg -l mirage /r2/data ---")
    output = await (await ws.execute("rg -l mirage /r2/data")).stdout_str()
    lines = output.strip().splitlines() if output.strip() else []
    print(f"  Files: {lines}")
    print(f"  Stats: {ops_summary()}")

    print("\n=== JQ QUERIES ===\n")

    print("--- jq .metadata ---")
    result = await ws.execute("jq .metadata /r2/data/example.json")
    print(f"  {(await result.stdout_str()).strip()}")

    print("\n--- jq: all team names (nested [] iterator) ---")
    result = await ws.execute(
        "jq \".departments[].teams[].name\" /r2/data/example.json")
    print(f"  {(await result.stdout_str()).strip()}")

    print("\n--- jq: all employee names ---")
    result = await ws.execute("jq \".departments[].teams[].members[].name\""
                              " /r2/data/example.json")
    print(f"  {(await result.stdout_str()).strip()}")

    print("\n--- jq: senior engineers on platform ---")
    result = await ws.execute("jq \".departments[0].teams[0].members"
                              " | map(select(.level == \\\"senior\\\"))"
                              " | map(.name)\" /r2/data/example.json")
    print(f"  {(await result.stdout_str()).strip()}")

    print("\n--- jq: all active project names ---")
    result = await ws.execute("jq \".departments[].teams[].projects"
                              " | map(select(.status == \\\"active\\\"))"
                              " | map(.name)\" /r2/data/example.json")
    print(f"  {(await result.stdout_str()).strip()}")

    print("\n--- jq: mirage project metrics ---")
    result = await ws.execute("jq .departments[0].teams[0].projects[0].metrics"
                              " /r2/data/example.json")
    print(f"  {(await result.stdout_str()).strip()}")

    print("\n--- jq: total budget ---")
    result = await ws.execute("jq .metadata.total_budget /r2/data/example.json"
                              )
    print(f"  {(await result.stdout_str()).strip()}")

    print("\n--- jq: compact JSON row ---")
    result = await ws.execute("jq -c .metadata /r2/data/example.json")
    print(f"  {(await result.stdout_str()).strip()}")

    print("\n--- jq: departments pretty json ---")
    result = await ws.execute("jq .departments /r2/data/example.json")
    parsed = json.loads(await result.stdout_str())
    print(f"  departments: {len(parsed)}")
    print(f"  first team: {parsed[0]['teams'][0]['name']}")

    # chmod/chown/touch never hit the R2 API: attrs land in the
    # workspace namespace (durable, snapshot-captured) and merge into
    # dispatch-level stat.
    print("=== metadata overlay on /r2/data/example.jsonl ===")
    meta_res = await ws.execute(
        'chmod 640 "/r2/data/example.jsonl"'
        ' && chown 500:dev "/r2/data/example.jsonl"'
        ' && touch -t 202601021530 "/r2/data/example.jsonl"')
    print(f"  chmod/chown/touch exit={meta_res.exit_code}")
    meta_st, _ = await ws.dispatch(
        "stat", PathSpec.from_str_path("/r2/data/example.jsonl"))
    print(f"  dispatch stat: mode={oct(meta_st.mode)[2:]} uid={meta_st.uid} "
          f"gid={meta_st.gid} mtime={meta_st.modified}")


if __name__ == "__main__":
    asyncio.run(main())
